File: src/leetcode/python/array/jump_game_55.py
# encoding='utf-8'

'''
   author:zhangyu
   date:2019.8.29
   description:找出是否能跳出这个数组
'''


def get_judge_jump_arr(arr):
    distance = 0
    length = len(arr)
    for index in range(1, length + 1):
        distance = max(distance, arr[index - 1])
        if distance == 0 and index < len(arr):
            return False
        if distance > len(arr) - index:
            return True
    return True

def get_judge_jump_arr2(arr):
    max = 0
    length = len(arr)
    for index in range(1, length):
        if arr[index - 1] == 0 and index > arr[max] + max:
            return False
        if arr[max] - (index - max) < arr[index]:
            max = index
    return True

def get_judge_jump_arr(arr):
    max = 0
    length = len(arr)
    for index in range(1, length):
        if arr[index - 1] == 0 and index > arr[max] + max:
            return False
        if arr[max] - (index - max) < arr[index]:
            max = index
    return True
if __name__ == '__main__':
    arr = [2, 3, 1, 1, 4]
    flag = get_judge_jump_arr(arr)
    print(flag)

File: src/leetcode/java/array/JumpGame55.java
package leetcode.java.array;

import org.junit.Test;

/**
 * @author zhangyu
 **/


public class JumpGame55 {
    @Test
    public void testJumpGame() {
        int[] nums = {2, 3, 1, 1, 4};
        boolean flag = canJump2(nums);
        System.out.println(flag);
    }

    /**
     * 能跳
     *
     * @param nums 数组
     * @return 布尔值
     */
    public boolean canJump2(int[] nums) {
        int n = nums.length;
        int range = 0;
        for (int i = 1; i <= n; i++) {
            range = Math.max(range - 1, nums[i - 1]);
            if (range == 0 && i < n)
                return false;
            if (range > n - 1)
                return true;
        }
        return true;
    }

    /**
     * 跳步游戏
     *
     * @param nums 数组
     * @return 布尔值
     */
    private boolean jumpGame(int[] nums) {
        int max = 0;
        for (int i = 1; i < nums.length; i++) {
            if (nums[i - 1] == 0 && i > max + nums[max]) {
                return false;
            }
            //max = nums[i] > nums[max] - (i - max) ? i : max;
            if (nums[i] > nums[max] - (i - max)) {
                max = i;
            }
        }
        return true;
    }
}
